[READY] Resolve completion item detailed_info on demand (where possible)

puremourning opened the pull request:
This goes with https://github.com/ycm-core/ycmd/pull/1473

In the client, the implementation is a bit fiddly, but boils down to this section of the vim help:

```
						*complete-popuphidden*
If the information for the popup is obtained asynchronously, use "popuphidden"
in 'completeopt'.  The info popup will then be initially hidden and
|popup_show()| must be called once it has been filled with the info.  This can
be done with a |CompleteChanged| autocommand, something like this: >
	set completeopt+=popuphidden
	au CompleteChanged * call UpdateCompleteInfo()
	func UpdateCompleteInfo()
	  " Cancel any pending info fetch
	  let item = v:event.completed_item
	  " Start fetching info for the item then call ShowCompleteInfo(info)
	endfunc
	func ShowCompleteInfo(info)
	  let id = popup_findinfo()
	  if id
	    call popup_settext(id, 'async info: ' .. a:info)
	    call popup_show(id)
	  endif
	endfunc
```

So:

* we can request the resolve on `CompleteChanged`
* we already store the extra_data in the completion user data, so we just pull out the `resolve` key from there
* when we have the resolved data, we call `popup_findinfo()` and then `popup_settext()` and `popup_show()` on that ID
* we re-use the `_latest_completion_request` so that we don't have outstanding resolve and completion requests
* we re-use the `request_data` from the completion request for efficiency and correctness (server will reject if the request data changes _at all_)
* we re-use the 'completion' poller, for the same reasons.

Simple, right ?

Well, there are some caveats:

* We only want to do this if Vim supports it
* We only want to do this if `completeopt` contains `popup` (we add `popuphidden` ourselves). Otherwise, we just break the preview window
* Some items will already be resolved, so we don't want to waste time re-resolving
* We should provide a way to turn off this async resolve in case the delay is too much or it's buggy. We use `g:ycm_max_num_candidates_to_resolve` for this with `-1` meaning no limit (the current behaviour). We force the value of this to 10 if we're going to do async resolve. This is a sort of cheap 'capabilities' implementation for this feature.


TODO:

* [x] Can we put `popuphidden` around the call to `complete()` like we do with `noselect` ? I tried it a while back and it didn't work.
* [x] Can we update the completion items in-place  ? I think we can't (text lock) and/or other reasons, but check and study vim code to see if enhacing vim is an option.
* [x] Write the tests!
* [x] Update README

For the record, none of tis works in neovim, unsurprisingly.

cc @habamax

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ycm-core/youcompleteme/3758)
<!-- Reviewable:end -->

puremourning reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758)* status: 0 of 2 LGTMs obtained

___
*[autoload/youcompleteme.vim, line 139 at r1](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MGzisDC0q9td13NIaRo:-MGzisDC0q9td13NIaRp:b-x8uy03) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7a00662bbe7c9e678c0cb53383b44433eb6a608e/autoload/youcompleteme.vim#L139)):*
> ```vim
>   " check, users on a recent vim and disabling preview by setting it to 0
>   " get `popup` added instead.
>   let s:force_preview_popup =
> ```

I moved this calculation as we need to use it to determine if we're going to resolve candidates and we'll also use it later for setting complete opt.

We need this here because we have to pass resolve_candidates to the YCM constructor.
___
*[autoload/youcompleteme.vim, line 246 at r1](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MGzj4Pw8uRsOgcYbSX0:-MGzj4Pw8uRsOgcYbSX1:b-eii59z) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7a00662bbe7c9e678c0cb53383b44433eb6a608e/autoload/youcompleteme.vim#L246)):*
> ```vim
>                  p.join( third_party_folder, 'requests-futures' ),
>                  p.join( third_party_folder, 'ycmd' ),
>                  p.join( third_party_folder, 'ycmd', 'third_party', 'cregex', 'regex_3' ),
> ```

this is junk remove it.

We deliberately _dont_ do this to prevent it crashing vim if the versions doing match client/server.
___
*[autoload/youcompleteme.vim, line 540 at r1](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MGzjE3NBAd8uC3pELqT:-MGzjE3NBAd8uC3pELqU:b-sn84l6) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7a00662bbe7c9e678c0cb53383b44433eb6a608e/autoload/youcompleteme.vim#L540)):*
> ```vim
> 
>   if s:resolve_completions
>     set completeopt+=popuphidden
> ```

we should check if this can be scoped to just the `complete` call. I'm not bothered if this stuff doesn't work with the legacy completion API
___
*[autoload/youcompleteme.vim, line 613 at r1](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MGzjOo50jNHfAase-ET:-MGzjOo50jNHfAase-EU:b-nkw8f0) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7a00662bbe7c9e678c0cb53383b44433eb6a608e/autoload/youcompleteme.vim#L613)):*
> ```vim
>   if py3eval( 'ycm_state.ResolveCompletionItem( vim.eval( "a:item" ) )' )
>     call s:StopPoller( s:pollers.completion )
>     " OK so this "works" but it resets the completion state each time. i think
> ```

remove this comment, it's old and untrue.
___
*[autoload/youcompleteme.vim, line 1062 at r1](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MGzjTGm4TP2nqk_Jnn0:-MGzjTGm4TP2nqk_Jnn1:b45ik2z) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7a00662bbe7c9e678c0cb53383b44433eb6a608e/autoload/youcompleteme.vim#L1062)):*
> ```vim
>   endif
> 
>   let completion =
> ```

it seems there's no way to update the list of completion candidates in-place here. at least, I don't _think_ so.

TODO: test what happens if you do `a:item.info = completion.info` here. Does that update the completion item in-place and prevent double-resolving ?
___
*[python/ycm/client/completion_request.py, line 72 at r1](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MGzjj1s9ZILcyvJj7MG:-MGzjj1s9ZILcyvJj7MH:b-oh0u52) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7a00662bbe7c9e678c0cb53383b44433eb6a608e/python/ycm/client/completion_request.py#L72)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
>     # FIXME: Do we really need to do this AdjustCandidateInsertionText ? I feel
>     # like Vim should do that for us
>     response[ 'completions' ] = base.AdjustCandidateInsertionText(
>         response[ 'completions' ] )
> ```
</details>


This call was moved from YouCompleteMe.GetCompletionResponse() because now that can return either a completion or a resolve response.

the comment is that it just seems odd that YCM does all this fix up on the text. now that we are sure to get the start_column right, it might be all redundant. I haven't studied what it does in detail, but it all seems kinda wonky.


<!-- Sent from Reviewable.io -->

bstaletic reviewed (changes requested):
> Reviewed 4 of 5 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758)* status: 0 of 2 LGTMs obtained (waiting on @puremourning)

___
*[autoload/youcompleteme.vim, line 540 at r1](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MGzjE3NBAd8uC3pELqT:-MH1lJLK7NAkXwMHS84x:bvd21sq) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7a00662bbe7c9e678c0cb53383b44433eb6a608e/autoload/youcompleteme.vim#L540)):*
> I'm not bothered if this stuff doesn't work with the legacy completion API

Agreed.
___
*[autoload/youcompleteme.vim, line 147 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1kTD979t77uVALfHt:-MH1kTD979t77uVALfHu:b3jv8od) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L147)):*
> ```vim
> 
>   " We can resolve completion items on demand if the user alaready set
>   " completeopt += popu, or if they asked us to do it for tham, and vim supports
> ```

s/popu/popup/
___
*[autoload/youcompleteme.vim, line 151 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1nZbj38LVlsSEjfoZ:-MH1nZbj38LVlsSEjfo_:btigeri) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L151)):*
> ```vim
>   let s:resolve_completions =
>         \ ( index( completeopt, 'popup' ) >= 0 || s:force_preview_popup ) &&
>         \ exists( '*popup_findinfo' )
> ```

Are we sure that this covers the existence of `completeopt+=popuphidden`?
___
*[autoload/youcompleteme.vim, line 153 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1khfl1qZkb_XhUedn:-MH1khfl1qZkb_XhUedo:b-vgvgzc) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L153)):*
> ```vim
>         \ exists( '*popup_findinfo' )
> 
>   let s:detail_completions =
> ```

What's the difference between `s:resolve_completions` and `s:detail_completions`?
___
*[autoload/youcompleteme.vim, line 1068 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1p5AR3Gk2bhd8b-la:-MH1p5AR3Gk2bhd8b-lb:b-sg1r9e) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L1068)):*
> ```vim
>     let s:pollers.completion.id = timer_start(
>           \ s:pollers.completion.wait_milliseconds,
>           \ function( 's:PollResolve', [ a:item ] ) )
> ```

Should this be `function( 's:PollResolve', a:item )`?

Inside `s:ResolveCompletionItem( item )`, `a:item` is a single item. It is passed to `s:PollResolve()` as `[ a:item ]`, thus when `s:PollResolve()` "calls" itself recursively, don't we end up with `[[item]]`?
___
*[autoload/youcompleteme.vim, line 1073 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1rZWb-98kBjUcuQTq:-MH1rZWb-98kBjUcuQTr:b-gyk6w2) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L1073)):*
> ```vim
> 
>   let completion =
>         \ py3eval( 'ycm_state.GetCompletionResponse()[ "completion" ]' )
> ```

This here actually returns a `ResolveCompletionRequest().Response()`, which is totally not obvious.

With this PR `ycm_state._latest_completion_request` may hold a `CompletionRequest` or a `ResolveCompletionItemRequest`.
___
*[autoload/youcompleteme.vim, line 1078 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1sc4H7dJSCP5rS4bs:-MH1sc4H7dJSCP5rS4bt:b-mr1oa2) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L1078)):*
> ```vim
>   endif
> 
>   call s:ShowInfoPopup( completion )
> ```

I'd call it `completion_item`.
___
*[autoload/youcompleteme.vim, line 1082 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1seN6BFi0LvWdEu8v:-MH1seN6BFi0LvWdEu8w:bulcr83) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L1082)):*
> ```vim
> endfunction
> 
> function! s:ShowInfoPopup( completion )
> ```

Again, `completion_item`.
___
*[autoload/youcompleteme.vim, line 1085 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1tC5O-yEWRI9emqrJ:-MH1tC5O-yEWRI9emqrK:b-p2m6tw) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L1085)):*
> ```vim
>   let id = popup_findinfo()
>   if id
>     call popup_settext( id, split( a:completion.info, '\n' ) )
> ```

You can use `a:completion.info->split( '\n' )`, though I don't know since which vim. So let's not change it. I'm not even sure `'asd'->split('s')` is better than `split('asd', 's')`.
___
*[python/ycm/youcompleteme.py, line 630 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1rAQc-_82JXU5yvB6:-MH1rAQd3G0s0PHw7vol:b-2049wp) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/python/ycm/youcompleteme.py#L630)):*
> ```Python
>       return False
> 
>     self._latest_completion_request = request
> ```

This was a surprise. `ResolveCompletionItem()` starts a `ResolveCompletionRequest()`. Might be worth a comment?
___
*[python/ycm/client/completion_request.py, line 31 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1mWSc-OfVD3zWBPkR:-MH1mWSc-OfVD3zWBPkS:b6abex0) ([raw file](https://github.com/ycm-core/youcompleteme/blob/37a149ee028bfb61a6e33dbfe30d7ce1343c9519/python/ycm/client/completion_request.py#L31)):*
> ```Python
> class CompletionRequest( BaseRequest ):
>   def __init__( self, request_data ):
>     super( CompletionRequest, self ).__init__()
> ```

How did I skip that?
___
*[python/ycm/client/completion_request.py, line 165 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1qSPf-Utk4-PAak1f:-MH1qSPf-Utk4-PAak1g:b-g2d5vc) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/python/ycm/client/completion_request.py#L165)):*
> ```Python
>     return None
>   try:
>     completion_extra_data = json.loads( item[ 'user_data' ] )
> ```

Since the next bump in the minimal required vim version will allow us to store `item[ 'user_data' ]` as an actual object, we'll be able to skip all this `json.loads` and `json.dumps` calls. Put a comment about it, so that it's easier to find.
___
*[python/ycm/client/completion_request.py, line 168 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1qpAuDmm4Jwmays1s:-MH1qpAuDmm4Jwmays1t:bwjm52k) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/python/ycm/client/completion_request.py#L168)):*
> ```Python
>   except KeyError:
>     return None
>   except json.JSONDecodeError:
> ```

You can put both exceptions on the same line.


<!-- Sent from Reviewable.io -->

puremourning reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758)* status: 0 of 2 LGTMs obtained (waiting on @bstaletic and @puremourning)

___
*[autoload/youcompleteme.vim, line 151 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1nZbj38LVlsSEjfoZ:-MH2on2OCfmdFE4YwSad:bhd58xj) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L151)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Are we sure that this covers the existence of `completeopt+=popuphidden`?
</blockquote></details>

it doesn't, but iirc popuphidden has no effect if popup isn't present. I'll check.
___
*[autoload/youcompleteme.vim, line 153 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1khfl1qZkb_XhUedn:-MH2o_4gDzHH_7MfUyj1:b-6eekp8) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L153)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

What's the difference between `s:resolve_completions` and `s:detail_completions`?
</blockquote></details>

s:detail_completions controls whether or not we should detail _any_ completions
s:resolve_completions controls whether or not we should use on-demand detailing.

Ultimately these both determine the value we use for `max_num_of_candidates_to_detail`
___
*[autoload/youcompleteme.vim, line 1068 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1p5AR3Gk2bhd8b-la:-MH2ovMj3TVlRmQRkd8q:b-ofruag) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L1068)):*
> Should this be function( 's:PollResolve', a:item )?

No, the argument to `function` is a list of arguments to make the `partial`.

So `function( 'X', [ a, b, c ] )` ultimately calls `X( a, b, c )`

so we're just calling `s:PollResolve( item, id )` (not `[ item ], id`)
___
*[autoload/youcompleteme.vim, line 1073 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1rZWb-98kBjUcuQTq:-MH2pI1uAmx3D6tLT-vw:b-jbjprr) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L1073)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

This here actually returns a `ResolveCompletionRequest().Response()`, which is totally not obvious.

With this PR `ycm_state._latest_completion_request` may hold a `CompletionRequest` or a `ResolveCompletionItemRequest`.
</blockquote></details>

That's right (as described in the PR description, this is intentional)
___
*[autoload/youcompleteme.vim, line 1085 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1tC5O-yEWRI9emqrJ:-MH2pQnuBDz5sGKAr7G1:beudwte) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L1085)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

You can use `a:completion.info->split( '\n' )`, though I don't know since which vim. So let's not change it. I'm not even sure `'asd'->split('s')` is better than `split('asd', 's')`.
</blockquote></details>

we can't use `->` methods as they don't work in neovim. (sigh)

if neovim implements the `popup_` APIs, then this should all work there too, I guess.
___
*[python/ycm/youcompleteme.py, line 630 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1rAQc-_82JXU5yvB6:-MH2phxlAbTOIT3lnBib:ba1lxek) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/python/ycm/youcompleteme.py#L630)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

This was a surprise. `ResolveCompletionItem()` starts a `ResolveCompletionRequest()`. Might be worth a comment?
</blockquote></details>

sure. I explained the reasoning int he PR description.
___
*[python/ycm/client/completion_request.py, line 168 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1qpAuDmm4Jwmays1s:-MH2pyd-Bc1eUXpDNkdc:b-qg52vu) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/python/ycm/client/completion_request.py#L168)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

You can put both exceptions on the same line.
</blockquote></details>

yeah, but I prefer to split them as I have the comment specifically in the JSONDecodeError branch. Otherwise, it's not obvious why we allow this.


<!-- Sent from Reviewable.io -->

bstaletic reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758)* status: 0 of 2 LGTMs obtained (waiting on @puremourning)

___
*[autoload/youcompleteme.vim, line 151 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1nZbj38LVlsSEjfoZ:-MH5GXyU2x40gHb6qX7J:bp9st44) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L151)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

it doesn't, but iirc popuphidden has no effect if popup isn't present. I'll check.
</blockquote></details>

If version of vim a user has doesn't know about `popuphidden`, then it will error with a backtrace. We've seen this, thankfully in CI, with `completeopt+=popup`.
___
*[autoload/youcompleteme.vim, line 153 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1khfl1qZkb_XhUedn:-MH5GpOT6Fh3gpwJHY5j:bokwqfi) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L153)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

s:detail_completions controls whether or not we should detail _any_ completions
s:resolve_completions controls whether or not we should use on-demand detailing.

Ultimately these both determine the value we use for `max_num_of_candidates_to_detail`
</blockquote></details>

Can we figure out better names. At one point during review, I did figure out the difference. Then I immediately forgot. These two sound the same to me.
___
*[autoload/youcompleteme.vim, line 1073 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1rZWb-98kBjUcuQTq:-MH5J-4cBxFagRLbmn84:b-lynvlh) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L1073)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

That's right (as described in the PR description, this is intentional)
</blockquote></details>

Do you think we should add a comment to call attention to that?
___
*[python/ycm/youcompleteme.py, line 630 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1rAQc-_82JXU5yvB6:-MH5JCjiAFVYJZ9U-FS7:bkdfrgy) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/python/ycm/youcompleteme.py#L630)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

sure. I explained the reasoning int he PR description.
</blockquote></details>

I agree with the reasoning. I should have also read the PR description more carefully. Do you think a comment would be useful?


<!-- Sent from Reviewable.io -->

bstaletic reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758)* status: 0 of 2 LGTMs obtained (waiting on @puremourning)

___
*[autoload/youcompleteme.vim, line 151 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1nZbj38LVlsSEjfoZ:-MH5NA4n1rydLkW945-U:b8paa68) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L151)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

If version of vim a user has doesn't know about `popuphidden`, then it will error with a backtrace. We've seen this, thankfully in CI, with `completeopt+=popup`.
</blockquote></details>

Right. This is broken between 8.1.1905 and 8.1.2192.


<!-- Sent from Reviewable.io -->

bstaletic reviewed (commented):
> Reviewed 9 of 9 files at r3.
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758)* status: 0 of 2 LGTMs obtained (waiting on @puremourning)

___
*[test/run_vim_tests, line 20 at r3](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHIMhTb1pwcSDm-o7ff:-MHIMhTc9QFxQv70i3hA:b1j5y8n) ([raw file](https://github.com/ycm-core/youcompleteme/blob/2c1a3287a9117284cea63abd7d6e0cae2a062d12/test/run_vim_tests#L20)):*
> ```
>   shift
>   shift
> fi
> ```

Interesting. What was your use case?


<!-- Sent from Reviewable.io -->

puremourning reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758)* status: 0 of 2 LGTMs obtained (waiting on @bstaletic and @puremourning)

___
*[autoload/youcompleteme.vim, line 151 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1nZbj38LVlsSEjfoZ:-MHIn8cHEKByFSPhh91d:b-896fix) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L151)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Right. This is broken between 8.1.1905 and 8.1.2192.
</blockquote></details>

Done.
___
*[test/run_vim_tests, line 20 at r3](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHIMhTb1pwcSDm-o7ff:-MHInAsp3eQVEAwsQ2cw:b-7thqzs) ([raw file](https://github.com/ycm-core/youcompleteme/blob/2c1a3287a9117284cea63abd7d6e0cae2a062d12/test/run_vim_tests#L20)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Interesting. What was your use case?
</blockquote></details>

I wanted to use a debug vim so I could attach to it to debug that popup_getinfo() issue


<!-- Sent from Reviewable.io -->

bstaletic reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758)* status: 0 of 2 LGTMs obtained (waiting on @puremourning)

___
*[autoload/youcompleteme.vim, line 151 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1nZbj38LVlsSEjfoZ:-MHKM5kI2PSD33ipbj6c:bt3eq0y) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L151)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Did you forget to push? I don't see any change.


<!-- Sent from Reviewable.io -->

puremourning reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758)* status: 0 of 2 LGTMs obtained (waiting on @bstaletic)

___
*[autoload/youcompleteme.vim, line 540 at r1](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MGzjE3NBAd8uC3pELqT:-MHbLZfgDN_nvp8tXxi9:b-qetpgr) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7a00662bbe7c9e678c0cb53383b44433eb6a608e/autoload/youcompleteme.vim#L540)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

> I'm not bothered if this stuff doesn't work with the legacy completion API

Agreed.
</blockquote></details>

there's no need to do this now - we dropped the legacy api
___
*[autoload/youcompleteme.vim, line 147 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1kTD979t77uVALfHt:-MHbDt547xtGJnx7NbA9:b-896fix) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L147)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

s/popu/popup/
</blockquote></details>

Done.
___
*[autoload/youcompleteme.vim, line 151 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1nZbj38LVlsSEjfoZ:-MHbDvKE9dAAo2HpfoFO:b-ohcwcb) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L151)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Did you forget to push? I don't see any change.
</blockquote></details>

Upped the min version to 8.1.2269
___
*[autoload/youcompleteme.vim, line 153 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1khfl1qZkb_XhUedn:-MHbLLPFC91G29w_91Ij:b-s5gb9g) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L153)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Can we figure out better names. At one point during review, I did figure out the difference. Then I immediately forgot. These two sound the same to me.
</blockquote></details>

Done.

I refactored this into an enum-like variable and added a load of commentary. I also made sure the 3 cases are all tested.
___
*[autoload/youcompleteme.vim, line 1073 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1rZWb-98kBjUcuQTq:-MHbMVwOB6wExocbRayN:b-896fix) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L1073)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Do you think we should add a comment to call attention to that?
</blockquote></details>

Done.
___
*[autoload/youcompleteme.vim, line 1078 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1sc4H7dJSCP5rS4bs:-MHbMdAt7GzuR9mZf36r:b-896fix) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L1078)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

I'd call it `completion_item`.
</blockquote></details>

Done.
___
*[autoload/youcompleteme.vim, line 1082 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1seN6BFi0LvWdEu8v:-MHbMe9T5sIFbtc50H7N:b-896fix) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/autoload/youcompleteme.vim#L1082)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Again, `completion_item`.
</blockquote></details>

Done.
___
*[python/ycm/youcompleteme.py, line 630 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1rAQc-_82JXU5yvB6:-MHbN7zR81fUC3PNL6B8:b-896fix) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/python/ycm/youcompleteme.py#L630)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

I agree with the reasoning. I should have also read the PR description more carefully. Do you think a comment would be useful?
</blockquote></details>

Done.
___
*[python/ycm/client/completion_request.py, line 165 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1qSPf-Utk4-PAak1f:-MHbNA4aEfSWXNEnakhk:b-v72hac) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/python/ycm/client/completion_request.py#L165)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Since the next bump in the minimal required vim version will allow us to store `item[ 'user_data' ]` as an actual object, we'll be able to skip all this `json.loads` and `json.dumps` calls. Put a comment about it, so that it's easier to find.
</blockquote></details>

Actually the bump isn't far enough, according to the comment - we're only bumping to 8.1.something, but this was changed in 8.2.something


<!-- Sent from Reviewable.io -->

bstaletic reviewed (changes requested):
Wow, I really like how much code this PR now drops.

> Reviewed 3 of 4 files at r4, 10 of 11 files at r5.
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-:-MHbWCJO-kIRVaiKABZI:b6s8dbt)* status: 0 of 2 LGTMs obtained (waiting on @puremourning)

___
*[README.md, line 17 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbTRiL88RSomisCXQV:-MHbTRiL88RSomisCXQW:b8h2kil) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/README.md#L17)):*
> ```Markdown
> 
> * [Legacy Python 2 support][legacy-py2]
> * [Legacy Vim support ][legacy-vim]
> ```

We've never before provided a legacy-vim branch. We simply bumped the required version to whatever was on Ubuntu latest-lts.

As for the python 2 branch, we wanted to support it for one year only.

In both cases I don't know if we should advertise these branches.
___
*[autoload/youcompleteme.vim, line 540 at r1](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MGzjE3NBAd8uC3pELqT:-MHbQ0Xn77_za1a_MgVE:b-oj9738) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7a00662bbe7c9e678c0cb53383b44433eb6a608e/autoload/youcompleteme.vim#L540)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

there's no need to do this now - we dropped the legacy api
</blockquote></details>

Just a reminder to mark this as resolved.
___
*[autoload/youcompleteme.vim, line 556 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbRuyhD1IJVFGx22uW:-MHbRuyhD1IJVFGx22uX:bjbhujk) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/autoload/youcompleteme.vim#L556)):*
> ```vim
> 
> 
> function! s:SetCompleteFunc()
> ```

This thing isn't used any more.
___
*[plugin/youcompleteme.vim, line 28 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbQu5y9Qk-0TTIf__C:-MHbQu5y9Qk-0TTIf__D:bbe9bvz) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/plugin/youcompleteme.vim#L28)):*
> ```vim
> 
> " NOTE: The minimum supported version is 8.1.2269, but neovim always reports as
> " v:version 800, but will largely work.
> ```

Let's complain to the neovim guys. Last time they fixed something this small within a day.
___
*[plugin/youcompleteme.vim, line 29 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbR5n_01Aw-wVZRIbc:-MHbR5n_01Aw-wVZRIbd:b-dm025o) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/plugin/youcompleteme.vim#L29)):*
> ```vim
> " NOTE: The minimum supported version is 8.1.2269, but neovim always reports as
> " v:version 800, but will largely work.
> let s:is_neovim = has( 'nvim' )
> ```

Is it worth caching this value? Or was it just because you find `s:is_neovim` more readable?
___
*[python/ycm/client/completion_request.py, line 165 at r2](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MH1qSPf-Utk4-PAak1f:-MHbQJmj7AnmLuaut54Y:b-2l0lgq) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7972832585e84ba88b0bf559bffbf9a19085971b/python/ycm/client/completion_request.py#L165)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

Actually the bump isn't far enough, according to the comment - we're only bumping to 8.1.something, but this was changed in 8.2.something
</blockquote></details>

That's a shame. The comment doesn't lie, because I've written it at the time I when I knew exactly what version we need for that.
___
*[test/lib/plugin/shared.vim, line 11 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbUypPCjf3DHiEYvRv:-MHbUypQ6ur_rOInT_-2:bq3imr8) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/test/lib/plugin/shared.vim#L11)):*
> ```vim
> " Run skip the current test if some expression returns true
> func SkipIf( expr, msg )
>   if type(a:expr) == v:t_func
> ```

Can we use `v:t_func`? If we can, then that check in autoload that says `type(option) == type('')` can be changed to `type(option) == v:t_string`, which I think is more readable.
___
*[test/completion_noresolve.test.vim, line 38 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbVieZC25e_NeZJPVj:-MHbVieZC25e_NeZJPVk:b-an8twd) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/test/completion_noresolve.test.vim#L38)):*
> ```vim
>   %bwipeout!
> endfunction
> ```

Extra empty line?


<!-- Sent from Reviewable.io -->

puremourning reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758)* status: 0 of 2 LGTMs obtained (waiting on @bstaletic and @puremourning)

___
*[autoload/youcompleteme.vim, line 556 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbRuyhD1IJVFGx22uW:-MHbcpJCBjZlGlbDtSxB:bl2tubg) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/autoload/youcompleteme.vim#L556)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

This thing isn't used any more.
</blockquote></details>

It is, just the name of the function is wrong now. Renamed it to s:EnableCompletingInCurrentBuffer()
___
*[plugin/youcompleteme.vim, line 28 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbQu5y9Qk-0TTIf__C:-MHbcx-S8LWpqhjsELlj:btiusqx) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/plugin/youcompleteme.vim#L28)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Let's complain to the neovim guys. Last time they fixed something this small within a day.
</blockquote></details>

I spoke to them and there's no real solution. v:version in Neovim is just not reliable in the sense that it diverged at 7.4 and doesn't pretend to be compatible with anything else. It could return 704 in v:version, but we would be in the same situation.
___
*[plugin/youcompleteme.vim, line 29 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbR5n_01Aw-wVZRIbc:-MHbd7A85mNQnYDqwvom:b9m18r0) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/plugin/youcompleteme.vim#L29)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Is it worth caching this value? Or was it just because you find `s:is_neovim` more readable?
</blockquote></details>

just more readable and allows adding the above comment to make it clear why we're doing it.
___
*[test/lib/plugin/shared.vim, line 11 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbUypPCjf3DHiEYvRv:-MHbdRKtCc0lA0991kLb:b3emfjp) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/test/lib/plugin/shared.vim#L11)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Can we use `v:t_func`? If we can, then that check in autoload that says `type(option) == type('')` can be changed to `type(option) == v:t_string`, which I think is more readable.
</blockquote></details>

yep it was added in 7.4.<something>. we can certainly use it in the tests. I'll change the other instances too; I actually thought this when I reviewed it but was meh about it.
___
*[test/completion_noresolve.test.vim, line 38 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbVieZC25e_NeZJPVj:-MHbdHtL499Zce3FKi4R:b-896fix) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/test/completion_noresolve.test.vim#L38)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Extra empty line?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

bstaletic reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758)* status: 0 of 2 LGTMs obtained (waiting on @puremourning)

___
*[autoload/youcompleteme.vim, line 556 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbRuyhD1IJVFGx22uW:-MHbf3Re7tWLfaq6rK_8:b-fwsdor) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/autoload/youcompleteme.vim#L556)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

It is, just the name of the function is wrong now. Renamed it to s:EnableCompletingInCurrentBuffer()
</blockquote></details>

Maybe I've done something wrong, but searching for the function name in reviewable didn't show any other results.

If it's still used, why don't we simply put `b:ycm_completing = 1` in the place of the call instead? A function that sets a single variable doesn't seem useful.
___
*[plugin/youcompleteme.vim, line 28 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbQu5y9Qk-0TTIf__C:-MHbfgmK49mU8CglyH2X:b-yhlfrj) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/plugin/youcompleteme.vim#L28)):*
> doesn't pretend to be compatible with anything else.

Hm...
___
*[plugin/youcompleteme.vim, line 29 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbR5n_01Aw-wVZRIbc:-MHbfpee51RBQaT2xvku:bev1kzq) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/plugin/youcompleteme.vim#L29)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

just more readable and allows adding the above comment to make it clear why we're doing it.
</blockquote></details>

I'm not sure it's more readable, since `has( 'nvim' )` is already well known, but that's fine.
___
*[test/lib/plugin/shared.vim, line 11 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbUypPCjf3DHiEYvRv:-MHbg9b62FO8gGuZm_Pk:b-dejziy) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/test/lib/plugin/shared.vim#L11)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

yep it was added in 7.4.<something>. we can certainly use it in the tests. I'll change the other instances too; I actually thought this when I reviewed it but was meh about it.
</blockquote></details>

Back in that PR, I was thinking that `v:t_string` was more readable, but then the docs mentioned that `type('')` would be needed for backwards compatibility. I was simply too lazy to check how new `v:t_string` actually is.
___
*[test/completion_noresolve.test.vim, line 38 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbVieZC25e_NeZJPVj:-MHbgmH3AbTwtEaHc2dp:b-uli6j8) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/test/completion_noresolve.test.vim#L38)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Did you forget to push?


<!-- Sent from Reviewable.io -->

bstaletic reviewed (commented):
The "Linux earliest Vim tests" are failing with:

>     Error: Invalid value for 'PROFILE_FILE...': Could not open file: .vim_profile: No such file or directory

> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-:-MHbhA478LAhm-utwnQQ:b-tda4ff)* status: 0 of 2 LGTMs obtained (waiting on @puremourning)



<!-- Sent from Reviewable.io -->

puremourning reviewed (commented):
Yeah I forgot to push the new docker image. fixed now.

> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-:-MHbmS9k0OxkZ-VkNiWZ:b-bp1pa9)* status: 0 of 2 LGTMs obtained (waiting on @puremourning)



<!-- Sent from Reviewable.io -->

puremourning reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758)* status: 0 of 2 LGTMs obtained (waiting on @bstaletic and @puremourning)

___
*[README.md, line 17 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbTRiL88RSomisCXQV:-MHbn5_l6lGbN2qd0LD7:b-896fix) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/README.md#L17)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

We've never before provided a legacy-vim branch. We simply bumped the required version to whatever was on Ubuntu latest-lts.

As for the python 2 branch, we wanted to support it for one year only.

In both cases I don't know if we should advertise these branches.
</blockquote></details>

Done.
___
*[autoload/youcompleteme.vim, line 556 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbRuyhD1IJVFGx22uW:-MHbnDnF5mfLMZxegtvN:b95zg63) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/autoload/youcompleteme.vim#L556)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Maybe I've done something wrong, but searching for the function name in reviewable didn't show any other results.

If it's still used, why don't we simply put `b:ycm_completing = 1` in the place of the call instead? A function that sets a single variable doesn't seem useful.
</blockquote></details>

I prefer the call-side semantic meaning associated with the new function name. It also means that a comment isn't needed to explain what b:ycm_completing means (it's implied by the method name).


<!-- Sent from Reviewable.io -->

puremourning reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758)* status: 0 of 2 LGTMs obtained (waiting on @bstaletic and @puremourning)

___
*[autoload/youcompleteme.vim, line 1062 at r1](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MGzjTGm4TP2nqk_Jnn0:-MHbrjLg-F9RXLMvny2e:b-nfhkip) ([raw file](https://github.com/ycm-core/youcompleteme/blob/7a00662bbe7c9e678c0cb53383b44433eb6a608e/autoload/youcompleteme.vim#L1062)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

it seems there's no way to update the list of completion candidates in-place here. at least, I don't _think_ so.

TODO: test what happens if you do `a:item.info = completion.info` here. Does that update the completion item in-place and prevent double-resolving ?
</blockquote></details>

it doesn't update the underlying completion data, and afaik there is no way to do that.


<!-- Sent from Reviewable.io -->

puremourning reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758)* status: 0 of 2 LGTMs obtained (waiting on @bstaletic)

___
*[test/completion_noresolve.test.vim, line 38 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbVieZC25e_NeZJPVj:-MHbsVIe9wThrr7uhGvu:b-jkgcga) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/test/completion_noresolve.test.vim#L38)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Did you forget to push?
</blockquote></details>

should be pushed now. 


<!-- Sent from Reviewable.io -->

bstaletic reviewed (approved):
What's up with codecov? It says than none of the new request related code is covered. Did we mess something up by adding another docker run?

> Reviewed 3 of 3 files at r6, 3 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-:-MHbxjWl59GhYB5u-UUF:bs1giy)* status: 0 of 2 LGTMs obtained

___
*[autoload/youcompleteme.vim, line 556 at r5](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbRuyhD1IJVFGx22uW:-MHboLIJ56hJrYTkU0kf:b-ox1a1v) ([raw file](https://github.com/ycm-core/youcompleteme/blob/0ecaa0077b7c7cb2227deecde133cdf1aefa1431/autoload/youcompleteme.vim#L556)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

I prefer the call-side semantic meaning associated with the new function name. It also means that a comment isn't needed to explain what b:ycm_completing means (it's implied by the method name).
</blockquote></details>

Fine with me.


<!-- Sent from Reviewable.io -->

puremourning reviewed (commented):
I'm not sure. it's as if the python part of the vim tests is not being uploaded (or perhaps recorded). I'll see if I can repro locally.

> Reviewed 5 of 9 files at r3, 3 of 4 files at r4, 5 of 11 files at r5, 2 of 3 files at r6, 3 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-:-MHc0NS21mdompzpD3iP:b-tm8ftz)* status: 0 of 2 LGTMs obtained



<!-- Sent from Reviewable.io -->

puremourning commented:
This is just awaiting ycmd PR to merge, then I need to update the ycmd submodule to the right SHA, then update README with any other ycmd changes since last update.

theli-ua commented on README.md:
```diff
@@ -1273,37 +1304,23 @@ setting one of the options. YCM will automatically pick the new values.
 
+YCM uses [rust-analyzer][] for Rust semantic completion.
+
+NOTE: Previously, YCM used [rls][] for ruest completion. This is no longer
```
>ruest

Typo

puremourning reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758)* status: 0 of 2 LGTMs obtained (waiting on @bstaletic, @puremourning, and @theli-ua)

___
*[README.md, line 1309 at r11](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHbBkoyozx6bD83Ei3F-r11-1309:-MHefp_Q1p5ebi32p48I:b-896fix) ([raw file](https://github.com/ycm-core/youcompleteme/blob/e0245c41ada3895adcfbde7e6fe3b91fe46015ee/README.md#L1309)):*
<details><summary><i>Previously, theli-ua (Anton Romanov) wrote…</i></summary><blockquote>

>ruest

Typo
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

bstaletic reviewed (changes requested):
Mostly looks good. @puremourning Mind generating the `docs/youcompleteme.txt` again?

> Reviewed 4 of 4 files at r8, 1 of 1 files at r9, 1 of 1 files at r10, 2 of 2 files at r12.
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-:-MHenl4tDi794FNmR1JT:bihtd5f)* status: 0 of 2 LGTMs obtained (waiting on @puremourning and @theli-ua)

___
*[README.md, line 1403 at r12](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHen6uo-DNvwX5gV2CE:-MHen6upCiGVMJJw7PIQ:b-ppax8y) ([raw file](https://github.com/ycm-core/youcompleteme/blob/202eab8ceaa5dc2745cf80724694a8ad4a70585c/README.md#L1403)):*
> ```Markdown
> ```
> 
> Each dictionary contains the following keys:
> ```

You skipped `capabilities`.
___
*[python/ycm/base.py, line 34 at r8](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHc7y4YDiEYVgERwE7V:-MHc7y4Z1g2VL1tlf0HS:b9xk6d2) ([raw file](https://github.com/ycm-core/youcompleteme/blob/fc93cff777beaa49e5bad3a5ce316df772117b0e/python/ycm/base.py#L34)):*
> ```Python
> 
>   # First load the default settings from ycmd. We do this to ensure that any
>   # client-side code that assumes all options are loaded (suc as the
> ```

s/suc/such
___
*[python/ycm/base.py, line 45 at r8](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHc8Cq8-4A9hMwNYicj:-MHc8Cq8-4A9hMwNYick:bk88efl) ([raw file](https://github.com/ycm-core/youcompleteme/blob/fc93cff777beaa49e5bad3a5ce316df772117b0e/python/ycm/base.py#L45)):*
> ```Python
>       user_options = json.load( defaults_file_handle )
> 
>   # Override the server defaults with any client-generated defaults
> ```

Override defaults with defaults and then override defaults with actual settings.

No issues here, just made me chuckle.
___
*[python/ycm/tests/completion_test.py, line 173 at r12](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHelpZs7n0hkzm09dov:-MHelpZs7n0hkzm09dow:b-mj3tuo) ([raw file](https://github.com/ycm-core/youcompleteme/blob/202eab8ceaa5dc2745cf80724694a8ad4a70585c/python/ycm/tests/completion_test.py#L173)):*
> ```Python
> 
> @YouCompleteMeInstance()
> @patch( 'ycm.client.base_request._logger', autospec = True )
> ```

What does `autospec` mean?


<!-- Sent from Reviewable.io -->

puremourning reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758)* status: 0 of 2 LGTMs obtained (waiting on @bstaletic and @theli-ua)

___
*[README.md, line 1403 at r12](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHen6uo-DNvwX5gV2CE:-MHepwa-70r86GBAiZ4a:b-896fix) ([raw file](https://github.com/ycm-core/youcompleteme/blob/202eab8ceaa5dc2745cf80724694a8ad4a70585c/README.md#L1403)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

You skipped `capabilities`.
</blockquote></details>

Done.
___
*[python/ycm/base.py, line 34 at r8](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHc7y4YDiEYVgERwE7V:-MHepxDl0ONTm_lAXZ_v:b-896fix) ([raw file](https://github.com/ycm-core/youcompleteme/blob/fc93cff777beaa49e5bad3a5ce316df772117b0e/python/ycm/base.py#L34)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

s/suc/such
</blockquote></details>

Done.
___
*[python/ycm/tests/completion_test.py, line 173 at r12](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHelpZs7n0hkzm09dov:-MHepr5W6H9fla4uQRBM:b-x5fsgb) ([raw file](https://github.com/ycm-core/youcompleteme/blob/202eab8ceaa5dc2745cf80724694a8ad4a70585c/python/ycm/tests/completion_test.py#L173)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

What does `autospec` mean?
</blockquote></details>

can't remember but I copied this from a test above which Michel wrote and he understood this stuff :)


<!-- Sent from Reviewable.io -->

bstaletic reviewed (changes requested):
> Reviewed 1 of 1 files at r13, 2 of 2 files at r14.
> Dismissed @theli-ua from a discussion.
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758)* status: 0 of 2 LGTMs obtained (waiting on @puremourning)

___
*[README.md, line 1421 at r14](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-MHf1JzZ-ppZXL3IKGF0:-MHeqQR197pHkv0GUmTn:b-jsuo5b) ([raw file](https://github.com/ycm-core/youcompleteme/blob/1395332a5f5f404845630a44c41d981295db07f9/README.md#L1421)):*
> ```Markdown
>   merged with the LSP client capabilities reported to the language server. This
>   can be used to enable or disable certain features, such as the support for
>   configuraiton sections (`workspace/configuration`).
> ```

Should we mention the related `section` thing from the extra conf?


<!-- Sent from Reviewable.io -->

bstaletic reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

I love it. Thanks for the PR!

> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-:-MHfI67C0D7sTHPUOmhX:b-kevybh)* status: 1 of 2 LGTMs obtained



<!-- Sent from Reviewable.io -->

puremourning reviewed (commented):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 2 of 4 files at r8, 1 of 1 files at r9, 1 of 1 files at r10, 1 of 2 files at r12, 2 of 2 files at r14.
> *[Reviewable](https://reviewable.io/reviews/ycm-core/youcompleteme/3758#-:-MHfIiMWAYDGRJTjDo5K:b-uwkowd)* status: :shipit: complete! 2 of 2 LGTMs obtained



<!-- Sent from Reviewable.io -->
